kotlin/builder/src/io/bazel/kotlin/builder/utils/jars/JarHelper.kt

```diff
@@ -0,0 +1,255 @@
+                outEntry.size = size
+                if (size == 0L) {
+                    outEntry.method = JarEntry.STORED
+                    outEntry.crc = 0
+                    putNextEntry(outEntry)
+                } else {
+                    outEntry.method = storageMethod
+                    if (storageMethod == JarEntry.STORED) {
+                        // ZipFile requires us to calculate the CRC-32 for any STORED entry.
+                        // It would be nicer to do this via DigestInputStream, but
+                        // the architecture of ZipOutputStream requires us to know the CRC-32
+                        // before we write the data to the stream.
+                        val bytes = Files.readAllBytes(path)
+                        val crc = CRC32()
+                        crc.update(bytes)
+                        outEntry.crc = crc.value
+                        putNextEntry(outEntry)
+                        write(bytes)
+                    } else {
+                        putNextEntry(outEntry)
+                        Files.copy(path, this)
+                    }
+                }
+                closeEntry()
+            }
+        }
+    }
+
+    /**
+     * Copies a a single entry into the jar. This variant differs from the other [copyEntry] in two ways. Firstly the
+     * jar contents are already loaded in memory and Secondly the [name] and [path] entries don't necessarily have a
+     * correspondence.
+     *
+     * @param path the path used to retrieve the timestamp in case normalize is disabled.
+     * @param bytes if this is null then the entry is a directory.
+     */
+    protected fun JarOutputStream.copyEntry(name: String, path: Path? = null, bytes: ByteArray = EMPTY_BYTEARRAY) {
+        val outEntry = JarEntry(name)
+        outEntry.time = when {
+            normalize -> normalizedTimestamp(name)
+            else -> Files.getLastModifiedTime(checkNotNull(path)).toMillis()
+        }
+        outEntry.size = bytes.size.toLong()
+
+        if (bytes.isEmpty()) {
+            outEntry.method = JarEntry.STORED
+            outEntry.crc = 0
+            putNextEntry(outEntry)
+        } else {
+            outEntry.method = storageMethod
+            if (storageMethod == JarEntry.STORED) {
+                val crc = CRC32()
+                crc.update(bytes)
+                outEntry.crc = crc.value
+                putNextEntry(outEntry)
+                write(bytes)
+            } else {
+                putNextEntry(outEntry)
+                write(bytes)
+            }
+        }
+        closeEntry()
+    }
+
+    companion object {
+        const val MANIFEST_DIR = "META-INF/"
+        const val MANIFEST_NAME = JarFile.MANIFEST_NAME
+        const val SERVICES_DIR = "META-INF/services/"
+        internal val EMPTY_BYTEARRAY = ByteArray(0)
+        /** Normalize timestamps.  */
+        val DEFAULT_TIMESTAMP = LocalDateTime.of(1980, 1, 1, 0, 0, 0)
```

Comment:
set to 1980 to align with the normalisation format of singlejar.

Reply:
There's nothing special about 2010-1-1. 1980-1-1 is fine in theory, but in practice using the earliest representable timestamp creates some opportunities for bugs (e.g. bazelbuild/bazel#4614).